lpd/views.py

```diff
@@ -1,27 +1,124 @@
+        qualitative_answers = json.loads(request.POST.get('qualitative_answers'))
+        quantitative_answers = json.loads(request.POST.get('quantitative_answers'))
+
+        print(user)
+        print(qualitative_answers)
+        print(quantitative_answers)
+
+        try:
+            self._process_qualitative_answers(user, qualitative_answers)
+            self._process_quantitative_answers(user, quantitative_answers)
+        except Exception as e:  # pylint: disable=broad-except
+            print(e)
+            response = JsonResponse({'message': 'Could not update learner answers.'}, status=500)
+        else:
+            response = JsonResponse({'message': 'Learner answers updated successfully.'})
+        return response
+
+    @classmethod
+    def _process_qualitative_answers(cls, user, qualitative_answers):
+        """
+        Create or update `QualitativeAnswer` for current learner, for each qualitative answer in request.
+        """
+        for qualitative_answer in qualitative_answers:
+            question_id = qualitative_answer.get('question_id')
+            question = QualitativeQuestion.objects.get(id=question_id)
+            text = qualitative_answer.get('answer_text')
+            answer, unused = QualitativeAnswer.objects.update_or_create(
+                learner=user,
+                question=question,
+                defaults=dict(
+                    text=text
+                ),
+            )
+            answer.save()
+
+    @classmethod
+    def _process_quantitative_answers(cls, user, quantitative_answers):
+        """
+        Create or update `QuantitativeAnswer` for current learner, for each quantitative answer in request.
+        """
+        for quantitative_answer in quantitative_answers:
+            question_type = quantitative_answer.get('question_type')
+            answer_option_id = quantitative_answer.get('answer_option_id')
+            answer_option = AnswerOption.objects.get(id=answer_option_id)
+            value = quantitative_answer.get('answer_option_value')
+            custom_input = quantitative_answer.get('answer_option_custom_input')
+            print(answer_option)
+            print(value)
+            if value is None:
+                if question_type == QuestionTypes.RANKING:
+                    value = RankingQuestion.unranked_option_value()
+                elif question_type == QuestionTypes.LIKERT:
+                    # If learner did not select a value for an answer to a Likert scale question,
+                    # we simply consider it unanswered.
+                    continue
+            answer, unused = QuantitativeAnswer.objects.update_or_create(
+                learner=user,
+                answer_option=answer_option,
+                defaults=dict(
+                    value=value
+                ),
+            )
+            if custom_input:
+                answer.custom_input = custom_input
```

Comment:
Could this bit be moved above `update_or_create` call in the following form:
```python
defaults = dict(value=value)
if custom_input:
   defaults[custom_input] = custom_input
```

?

I think that way we could avoid hitting db another time, no?

Reply:
@tomaszgy Definitely, that's a great idea. Applied that change in 82ab6fcd669707a6d83752961c2cf88f3119aea9.